src/main/java/nextstep/security/authorization/role/RoleHierarchyImpl.java

```diff
@@ -0,0 +1,106 @@
+package nextstep.security.authorization.role;
+
+import nextstep.security.SimpleGrantedAuthority;
+import nextstep.security.authorization.GrantedAuthority;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+public class RoleHierarchyImpl implements RoleHierarchy {
+
+    private final Map<String, Set<GrantedAuthority>> reachableAuthoritiesMap;
+
+    private RoleHierarchyImpl(Map<String, Set<GrantedAuthority>> reachableAuthoritiesMap) {
+        this.reachableAuthoritiesMap = reachableAuthoritiesMap;
+    }
+
+    public static RoleHierarchyImpl fromHierarchy(String hierarchy) {
+        Map<String, Set<GrantedAuthority>> reachableAuthoritiesMap = convertHierarchicalRoles(hierarchy);
+        return new RoleHierarchyImpl(reachableAuthoritiesMap);
+    }
+
+    public static Builder with() {
+        return new Builder();
+    }
+
+    private static Map<String, Set<GrantedAuthority>> convertHierarchicalRoles(String hierarchy) {
+
+        Map<String, Set<GrantedAuthority>> hierarchicalMap = new HashMap<>();
+        for (String line : hierarchy.split("\n")) {
+            String[] roles = line.trim().split("\\s*>\\s*");
+
+            String parent = roles[0].trim();
+            String child = roles[1].trim();
+
+            hierarchicalMap.computeIfAbsent(parent, k -> new HashSet<>())
+                    .add(new SimpleGrantedAuthority(child));
+        }
+
+        return hierarchicalMap;
+    }
+
+    @Override
+    public Collection<GrantedAuthority> getReachableGrantedAuthorities(
+            Collection<GrantedAuthority> authorities) {
+
+        Set<GrantedAuthority> result = new HashSet<>();
+        Set<GrantedAuthority> processed = new HashSet<>();
+        Queue<GrantedAuthority> queue = new LinkedList<>(authorities);
```

Comment:
```suggestion
        Queue<GrantedAuthority> queue = new ArrayDeque<>(authorities);
```

BFS 알고리즘을 활용하여 잘 구현해 주셨네요 👍 
알고리즘에 맞는 자료구조로 바꾼다면 성능이 더 올라갈 것 같습니다!